src/ui/public/state_management/state.js

```diff
@@ -1,59 +1,94 @@
 import _ from 'lodash';
+import angular from 'angular';
 import rison from 'rison-node';
 import applyDiff from 'ui/utils/diff_object';
 import qs from 'ui/utils/query_string';
 import EventsProvider from 'ui/events';
 import Notifier from 'ui/notify/notifier';
 import KbnUrlProvider from 'ui/url';
 
-const notify = new Notifier();
-export default function StateProvider(Private, $rootScope, $location) {
+import {
+  createStateHash,
+  hashedItemStoreSingleton,
+  isStateHash,
+} from './state_storage';
+
+export default function StateProvider(Private, $rootScope, $location, config) {
   const Events = Private(EventsProvider);
 
   _.class(State).inherits(Events);
-  function State(urlParam, defaults) {
+  function State(
+    urlParam,
+    defaults,
+    hashedItemStore = hashedItemStoreSingleton,
+    notifier = new Notifier()
+  ) {
     State.Super.call(this);
 
-    let self = this;
-    self.setDefaults(defaults);
-    self._urlParam = urlParam || '_s';
+    this.setDefaults(defaults);
+    this._urlParam = urlParam || '_s';
+    this._notifier = notifier;
+    this._hashedItemStore = hashedItemStore;
 
     // When the URL updates we need to fetch the values from the URL
-    self._cleanUpListeners = _.partial(_.callEach, [
+    this._cleanUpListeners = _.partial(_.callEach, [
       // partial route update, no app reload
-      $rootScope.$on('$routeUpdate', function () {
-        self.fetch();
+      $rootScope.$on('$routeUpdate', () => {
+        this.fetch();
       }),
 
       // beginning of full route update, new app will be initialized before
       // $routeChangeSuccess or $routeChangeError
-      $rootScope.$on('$routeChangeStart', function () {
-        if (!self._persistAcrossApps) {
-          self.destroy();
+      $rootScope.$on('$routeChangeStart', () => {
+        if (!this._persistAcrossApps) {
+          this.destroy();
         }
       }),
 
-      $rootScope.$on('$routeChangeSuccess', function () {
-        if (self._persistAcrossApps) {
-          self.fetch();
+      $rootScope.$on('$routeChangeSuccess', () => {
+        if (this._persistAcrossApps) {
+          this.fetch();
         }
       })
     ]);
 
     // Initialize the State with fetch
-    self.fetch();
+    this.fetch();
   }
 
   State.prototype._readFromURL = function () {
-    let search = $location.search();
+    const search = $location.search();
+    const urlVal = search[this._urlParam];
+
+    if (!urlVal) {
+      return null;
+    }
+
+    if (isStateHash(urlVal)) {
+      return this._parseQueryParamValue(urlVal);
+    }
+
+    let risonEncoded;
+    let unableToParse;
     try {
-      return search[this._urlParam] ? rison.decode(search[this._urlParam]) : null;
+      risonEncoded = rison.decode(urlVal);
```

Comment:
below is minor.

I like the error is trapped very narrowly, but passing forward state (`unableToParse`) to then unwind the error handling is not necessary IMHO. That'd also avoid the multiple ifs and return points later on.

if the rison.decode function is guaranteed to return a value, perhaps consider something like:

    let risonEncoded;
    try {
      risonEncoded = rison.decode(urlVal);
    } catch (e) {
      this._notifier.error('Unable to parse URL');
      search[this._urlParam] = this.toQueryParam(this._defaults);
      $location.search(search).replace();
      return null;
    }

    search[this._urlParam] = this.toQueryParam(risonEncoded);
    $location.search(search).replace();
    return risonEncoded;